### PR TITLE
Replace packaging with packvers #108

### PR DESCRIPTION
**Pull request by swastkk**
Packvers added in resolution.py and changed the version of pip-requirments-parser too.
for the https://github.com/nexB/scancode.io/issues/576 and https://github.com/nexB/python-inspector/issues/108 issue as well.
Signed-off-by: Swastik Sharma <swastkk@gmail.com>

### CODE REVIEW

**Comment by swastkk** [comment]
Getting this error while using `make test` command.
```sh 
=========================================== warnings summary ===========================================
tests/test_api.py::test_api_with_requirement_file
tests/test_cli.py::test_pdt_output_with_pinned_requirements
tests/test_cli.py::test_pdt_output_with_frozen_requirements
tests/test_cli.py::test_cli_with_pinned_requirements_file
tests/test_cli.py::test_cli_with_setup_py
  /home/swastik/Desktop/Projects/python-inspector/venv/lib/python3.10/site-packages/packaging/specifiers.py:255: DeprecationWarning: Creating a LegacyVersion has been deprecated and will be removed in the next major release
    warnings.warn(

tests/test_resolution.py::test_without_supported_wheels
tests/test_resolution.py::test_without_supported_wheels
  /home/swastik/Desktop/Projects/python-inspector/venv/lib/python3.10/site-packages/packaging/version.py:111: DeprecationWarning: Creating a LegacyVersion has been deprecated and will be removed in the next major release
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================================= short test summary info ========================================
FAILED tests/test_api.py::test_api_with_specifier - TypeError: expected string or bytes-like object
FAILED tests/test_api.py::test_api_with_specifier_pdt - TypeError: expected string or bytes-like object
FAILED tests/test_api.py::test_api_with_requirement_file - TypeError: expected string or bytes-like o...
FAILED tests/test_api.py::test_api_with_prefer_source - TypeError: expected string or bytes-like object
FAILED tests/test_api.py::test_api_with_recursive_requirement_file - TypeError: expected string or by...
FAILED tests/test_api.py::test_api_with_python_311 - TypeError: expected string or bytes-like object
FAILED tests/test_cli.py::test_cli_with_default_urls - AssertionError:
FAILED tests/test_cli.py::test_pdt_output - AssertionError:
FAILED tests/test_cli.py::test_pdt_output_with_pinned_requirements - AssertionError:
FAILED tests/test_cli.py::test_pdt_output_with_frozen_requirements - AssertionError:
FAILED tests/test_cli.py::test_cli_with_single_index_url - AssertionError:
FAILED tests/test_cli.py::test_cli_with_single_index_url_except_pypi_simple - AssertionError:
FAILED tests/test_cli.py::test_cli_with_multiple_index_url_and_tilde_req - AssertionError:
FAILED tests/test_cli.py::test_cli_with_environment_marker_and_complex_ranges - AssertionError:
FAILED tests/test_cli.py::test_cli_with_azure_devops_with_python_310 - AssertionError:
FAILED tests/test_cli.py::test_cli_with_azure_devops_with_python_38 - AssertionError:
FAILED tests/test_cli.py::test_cli_with_multiple_index_url_and_tilde_req_with_max_rounds - AssertionE...
FAILED tests/test_cli.py::test_cli_with_multiple_index_url_and_tilde_req_and_netrc_file_without_matching_url
FAILED tests/test_cli.py::test_cli_with_prefer_source - AssertionError:
FAILED tests/test_cli.py::test_cli_with_pinned_requirements_file - AssertionError:
FAILED tests/test_cli.py::test_cli_with_insecure_option - AssertionError:
FAILED tests/test_cli.py::test_cli_with_insecure_option_testpkh - AssertionError:
FAILED tests/test_cli.py::test_cli_with_insecure_option_testrdflib - AssertionError:
FAILED tests/test_cli.py::test_cli_with_setup_py - AssertionError:
FAILED tests/test_resolution.py::test_get_resolved_dependencies_with_flask_and_python_310 - TypeError...
FAILED tests/test_resolution.py::test_get_resolved_dependencies_with_flask_and_python_310_windows - T...
FAILED tests/test_resolution.py::test_get_resolved_dependencies_with_flask_and_python_36 - TypeError:...
FAILED tests/test_resolution.py::test_get_resolved_dependencies_with_tilde_requirement_using_json_api
FAILED tests/test_resolution.py::test_without_supported_wheels - TypeError: expected string or bytes-...
============================== 29 failed, 94 passed, 7 warnings in 15.67s ==============================
make: *** [Makefile:48: test] Error 1
```

**Comment by swastkk** [comment]
After analysing I found that due to that change in Packvers, we need to update our https://github.com/nexB/dparse2 for packvers and packaging change in the release and https://github.com/nexB/pip-requirements-parser to in the release , after that this PR will have all Test pass 
Signed-off-by: swastik <swastkk@gmail.com>

File: etc/scripts/utils_dejacode.py [hunk 1]
```diff
@@ -15,7 +15,7 @@
 import requests
 import saneyaml
 
-from packaging import version as packaging_version
+from packvers import version as packaging_version
```
Comment (pombredanne) [inline etc/scripts/utils_dejacode.py:18]: etc/scripts comes from the https://github.com/nexB/skeleton and should be modified instead there

File: src/_packagedcode/pypi.py [hunk 2]
```diff
@@ -22,15 +22,15 @@
 import zipfile
 
 import dparse2
-import packaging
```
Comment (pombredanne) [inline src/_packagedcode/pypi.py:25]: This pypi.py modules comes originally from ScanCode toolkit... Can you check how the modifications were done there and have the same modifications?

File: setup.cfg [hunk 3]
```diff
@@ -69,7 +69,8 @@ install_requires =
     tinynetrc >= 1.3.1
     toml >= 0.10.0
     mock >= 3.0.5
-
+    # packaging >= 21.3
```
Comment (pombredanne) [inline setup.cfg:72]: We still use packaging in pip-requirements-parser and you cannot remove it. Only add packvers. See in ScanCode toolkit the proper minimal versions to use  here and to pin in the requirements.txt file. You also need to bump dparse2 and pip-requirements-parser here and in requirements.txt

**Review by pombredanne — CHANGES_REQUESTED** [review]
Thank you ++ !

"Packvers added" may not be the best name for this PR. What about "Replace packaging with packvers" instead?

Also on the commit message style, your use of things such as  "packaging-to-packvers-done-in-resolution.py" may not be something we typically use... look for other commit messages for inspiration and also at https://aboutcode.readthedocs.io/en/latest/contributing/writing_good_commit_messages.html

See also the other review comments I posted for your kind consideration.

**Comment by swastkk** [comment]
> Thank you ++ !
> 
> "Packvers added" may not be the best name for this PR. What about "Replace packaging with packvers" instead?
> 
> Also on the commit message style, your use of things such as  "packaging-to-packvers-done-in-resolution.py" may not be something we typically use... look for other commit messages for inspiration and also at https://aboutcode.readthedocs.io/en/latest/contributing/writing_good_commit_messages.html
> 
> See also the other review comments I posted for your kind consideration.
> 
> 
> 
> 
> 

Okay sure will change accordingly : )

File: etc/scripts/utils_dejacode.py (hunk 1 above)
Comment (swastkk) [reply etc/scripts/utils_dejacode.py:18]: Modification done there , please merge that PR https://github.com/nexB/skeleton/pull/71 :")

File: setup.cfg (hunk 3 above)
Comment (swastkk) [reply setup.cfg]: Done in https://github.com/nexB/python-inspector/pull/109/commits/d542835d9480e86cd03e493a6f4badd345534721 and https://github.com/nexB/python-inspector/pull/109/commits/06ea29bf33efa7baed2c158205bbcb3cddad50e0

File: etc/scripts/utils_pip_compatibility_tags.py [hunk 4]
```diff
@@ -27,7 +27,7 @@
 
 import re
 
-from packaging.tags import (
+from packvers.tags import (
```
Comment (pombredanne) [inline etc/scripts/utils_pip_compatibility_tags.py:30]: Let's leave this to a future merge of the skeleton. Please remove this change

File: etc/scripts/utils_thirdparty.py [hunk 5]
```diff
@@ -28,8 +28,8 @@
 from commoncode import fileutils
 from commoncode.hash import multi_checksums
 from commoncode.text import python_safe_name
-from packaging import tags as packaging_tags
-from packaging import version as packaging_version
+from packvers import tags as packaging_tags
+from packvers import version as packaging_version
```
Comment (pombredanne) [inline etc/scripts/utils_thirdparty.py:32]: Let's leave this to a future merge of the skeleton. Please remove these changes

File: src/python_inspector/utils_pypi.py [hunk 6]
```diff
@@ -107,8 +107,12 @@
     "39": "3.9",
     "310": "3.10",
     "27": "2.7",
+    "311": "3.11",
 }
 
+valid_python_versions = list(PYTHON_DOT_VERSIONS_BY_VER.keys())
+valid_python_versions.extend([dot_ver for pyver, dot_ver in PYTHON_DOT_VERSIONS_BY_VER.items()])
+
```
Comment (pombredanne) [inline src/python_inspector/utils_pypi.py:115]: Why did you make this change in these two lines?

File: CHANGELOG.rst [hunk 7]
```diff
@@ -6,6 +6,10 @@ v0.9.3
 ------
 
 - Add support for recursive requirements.
+- Add python 3.11 as a valid python version in choices.
+- Operating system and python version are now required fields in CLI.
+- Add dot versions (3.6, 3.7, 3.8, 3.9, 3.10, 3.11, 2.7) with
```
Comment (pombredanne) [inline CHANGELOG.rst:11]: I am not sure I agree with this change. 
I get that using versions without dot in the CLI is not a great UX, but supporting both is likely a source of confusion and inconsitencies.

File: src/python_inspector/api.py [hunk 8]
```diff
@@ -103,9 +103,6 @@ def resolve_dependencies(
             f"Must be one of: {', '.join(PLATFORMS_BY_OS.keys())}"
         )
 
-    valid_python_versions = list(PYTHON_DOT_VERSIONS_BY_VER.keys())
```
Comment (pombredanne) [inline src/python_inspector/api.py:106]: This would be an API breaking change and I am not sure I get the value for this

**Review by pombredanne — CHANGES_REQUESTED** [review]
Thank you ... here are some review comments

File: src/python_inspector/utils_pypi.py (hunk 6 above)
Comment (swastkk) [reply src/python_inspector/utils_pypi.py:115]: Actually I didn't made these changes, these changes automatically gets added to my PR after I ran git pull command and changed the head to different position, but will remove these for now.

File: src/python_inspector/api.py (hunk 8 above)
Comment (swastkk) [reply src/python_inspector/api.py]: https://github.com/nexB/python-inspector/pull/109/commits/3a66ca2e30d8fd6e668219317723843f138145ab Removed this stuff which were not part of this PR

File: CHANGELOG.rst (hunk 7 above)
Comment (swastkk) [reply CHANGELOG.rst]: https://github.com/nexB/python-inspector/commit/3a66ca2e30d8fd6e668219317723843f138145ab Removed this stuff which were not part of this PR

File: src/python_inspector/utils_pypi.py (hunk 6 above)
Comment (swastkk) [reply src/python_inspector/utils_pypi.py]: https://github.com/nexB/python-inspector/pull/109/commits/5a2ceb2e7948f0d9d8e6f564a72c4dc7b55dcb1f Removed this stuff which was not part of this PR

**Review by pombredanne — CHANGES_REQUESTED** [review]
Thank you for the updates. Do you mind to squash your commits in a single commit with a good commit message?

**Review by pombredanne — COMMENTED** [review]
LGTM and ready to merge once squashed!

**Comment by swastkk** [comment]
> LGTM and ready to merge once squashed!

Done ✔️, can merge now :")

**Comment by pombredanne** [comment]
@swastkk you did not squash it all in one commit, you still have 15 commits... you squash and force push. https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History

**Comment by swastkk** [comment]
> @swastkk you did not squash it all in one commit, you still have 15 commits... you squash and force push. https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History

Is it okay now??

**Review by pombredanne — CHANGES_REQUESTED** [review]
Thanks!
Can you make one last change and amend your commit?

We want to subject to be using imperative style (e.g., Replace not Replaced) and to be subject line must be under 50 characters and ew extra things to review at https://aboutcode.readthedocs.io/en/latest/contributing/writing_good_commit_messages.html

**Comment by swastkk** [comment]
> Thanks! Can you make one last change and amend your commit?
> 
> We want to subject to be using imperative style (e.g., Replace not Replaced) and to be subject line must be under 50 characters and ew extra things to review at https://aboutcode.readthedocs.io/en/latest/contributing/writing_good_commit_messages.html

Now?

File: src/python_inspector/api.py [hunk 9]
```diff
@@ -103,6 +103,9 @@ def resolve_dependencies(
             f"Must be one of: {', '.join(PLATFORMS_BY_OS.keys())}"
         )
 
+    valid_python_versions = list(PYTHON_DOT_VERSIONS_BY_VER.keys())
+    valid_python_versions.extend([dot_ver for pyver, dot_ver in PYTHON_DOT_VERSIONS_BY_VER.items()])
```
Comment (pombredanne) [inline src/python_inspector/api.py:107]: Remove this

File: src/python_inspector/resolve_cli.py [hunk 10]
```diff
@@ -70,10 +70,9 @@ def print_version(ctx, param, value):
     "-p",
     "--python-version",
     "python_version",
-    type=click.Choice(utils_pypi.valid_python_versions),
+    type=click.Choice(utils_pypi.PYTHON_VERSIONS),
     metavar="PYVER",
     show_default=True,
-    required=True,
```
Comment (pombredanne) [inline src/python_inspector/resolve_cli.py:76]: Remove this

File: src/python_inspector/resolve_cli.py [hunk 11]
```diff
@@ -83,7 +82,6 @@ def print_version(ctx, param, value):
     type=click.Choice(utils_pypi.PLATFORMS_BY_OS),
     metavar="OS",
     show_default=True,
-    required=True,
```
Comment (pombredanne) [inline src/python_inspector/resolve_cli.py:86]: Remove this change

**Review by pombredanne — CHANGES_REQUESTED** [review]
Thanks!
"Replace packaging with packvers and bump packages version"  is still at 57 chars... can you do something more or less like this instead:

```
Replace packaging library with packvers #108

Work around changes in packaging by replacing it with packvers
to avoid issues with missing packaging.version.LegacyVersion.

Also bump the versions of dparse2 and pip-requirements-parser
with new versions that are not subject the LegcacyVersion issue.

Adjust the code and tests accordingly.

Reference: https://github.com/nexB/python-inspector/issues/108
Reference: https://github.com/pypa/packaging/issues/530
Signed-off-by: swastik <swastkk@gmail.com>
```

And please remove all changes you made that are not strictly about this... I am not sure what happened but you end up with unrelated changes. Do not hesitate to ask for help on the Gitter chat!

Thanks again for your preseverance.

File: src/python_inspector/api.py [hunk 12]
```diff
@@ -40,8 +40,8 @@
 from python_inspector.resolution import get_requirements_from_python_manifest
 from python_inspector.utils_pypi import PLATFORMS_BY_OS
 from python_inspector.utils_pypi import PYPI_SIMPLE_URL
+from python_inspector.utils_pypi import PYTHON_DOT_VERSIONS_BY_VER
```
[the remark this answers is not in the document]
Comment (swastkk) [reply src/python_inspector/api.py:43]: Actually during squashing, it squashed  the already merged commits which @TG1999 did on main branch, ig these will not affect the working , isn't it?|

File: src/python_inspector/resolve_cli.py (hunk 10 above)
Comment (swastkk) [reply src/python_inspector/resolve_cli.py]: Done

File: src/python_inspector/resolve_cli.py (hunk 11 above)
Comment (swastkk) [reply src/python_inspector/resolve_cli.py]: done

File: src/python_inspector/resolve_cli.py [hunk 13]
```diff
@@ -70,10 +70,9 @@ def print_version(ctx, param, value):
     "-p",
     "--python-version",
     "python_version",
-    type=click.Choice(utils_pypi.valid_python_versions),
+    type=click.Choice(utils_pypi.PYTHON_VERSIONS),
```
[the remark this answers is not in the document]
Comment (swastkk) [reply src/python_inspector/resolve_cli.py]: Done too

File: src/python_inspector/api.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (swastkk) [reply src/python_inspector/api.py]: Sorry actually was merging problem, now merged right commits :")

File: src/_packagedcode/pypi.py [hunk 14]
```diff
@@ -22,15 +22,14 @@
...
+import packvers as packaging
 import pip_requirements_parser
-import pkginfo2
```
Comment (pombredanne) [inline src/_packagedcode/pypi.py:27]: Why did you remove thin import? Please reinstate

**Review by pombredanne — CHANGES_REQUESTED** [review]
Thanks for the update. One last nit before we can merge.

File: src/_packagedcode/pypi.py (hunk 14 above)
Comment (swastkk) [reply src/_packagedcode/pypi.py:27]: Added : )

Comment (pombredanne) [reply src/_packagedcode/pypi.py:27]: But you moved it to a spot that's not in the correct sort order and should be right after the `import pip_requirements_parser`

Comment (swastkk) [reply src/_packagedcode/pypi.py:27]: Done with the order :") , but wanted to know the reason.

Comment (pombredanne) [reply src/_packagedcode/pypi.py:27]: we sort package imports like isort would do

Comment (pombredanne) [reply src/_packagedcode/pypi.py:27]: Please just keep this this way:
```
import pip_requirements_parser
import pkginfo2
```

Comment (swastkk) [reply src/_packagedcode/pypi.py:27]: Ohk got it ✨

File: setup.cfg [hunk 15]
```diff
@@ -58,18 +58,19 @@ install_requires =
...
-    pip-requirements-parser >= 31.2.0
+    pip-requirements-parser >= 32.0.1
     requests >= 2.18.0
     resolvelib >= 0.8.1
     saneyaml >= 0.5.2
     tinynetrc >= 1.3.1
     toml >= 0.10.0
     mock >= 3.0.5
-
+    packaging >= 21.0.0
```
Comment (pombredanne) [inline setup.cfg:72]: we no longer have a direct dependency on packaging, so do not add this.

**Review by pombredanne — CHANGES_REQUESTED** [review]
Here are a few last nits.

File: src/_packagedcode/pypi.py [hunk 16]
```diff
@@ -20,18 +20,16 @@
...
+from packvers.specifiers import SpecifierSet
 from packageurl import PackageURL
-from packaging import markers
-from packaging.requirements import Requirement
-from packaging.utils import canonicalize_name
-
+from packvers import markers
+from packvers.requirements import Requirement
+from packvers.utils import canonicalize_name
+import pkginfo2
```
Comment (pombredanne) [inline src/_packagedcode/pypi.py:32]: See other comment. Do not move this here, leave it above.

Comment (swastkk) [reply src/_packagedcode/pypi.py]: Done check now.

**Review by pombredanne — APPROVED** [review]
All good .... Thank you ++

**Pull request merged**